### ISSUE TITLE
Migration tool rerun is slower than first run

### ISSUE THREAD

**Issue by srawlins** [dart-lang/sdk#44086]
TL;DR We run `InfoBuilder.explainMigration` _twice_ during a rerun, which I think builds up a huge graph of data. Additionally, it seems like even running `explainMigration` just the first time during a rerun takes much longer than the first time, like 3x to 8x as long.

As an anecdotal test, I deleted many test files from the analyzer code base, to reduce time, then ran `dart migrate`. I timed the following:

### First run:

1. `_FixCodeProcessor.runFirstPhase`, (the `NonNullableFix.prepareUnit`s), **22 seconds**
2. `_FixCodeProcessor.runLaterPhases`, (the `NonNullableFix.processUnit`s), **22 seconds**
3. `_FixCodeProcessor.runLaterPhases`, (the `NonNullableFix.finalizeUnit`s), **23 seconds**
4. `_FixCodeProcessor.runLaterPhases`, (`NonNullableFix.finish` ,including `InfoBuilder.explainMigration`), **41 seconds**

### First "Rerun" in the UI:

1. `_FixCodeProcessor.runFirstPhase`, (the `NonNullableFix.prepareUnit`s), **13 seconds**
2. `_FixCodeProcessor.runLaterPhases`, (the `NonNullableFix.processUnit`s), **21 seconds**
3. `_FixCodeProcessor.runLaterPhases`, (the `NonNullableFix.finalizeUnit`s), **28 seconds**
4. `_FixCodeProcessor.runLaterPhases`, (`NonNullableFix.finish` ,including `InfoBuilder.explainMigration`), **146 seconds**
5. `NonNullableFix.rerun`, (`MigrationState.refresh`, which also calls `InfoBuilder.explainMigration`), **200 seconds**

### Second "Rerun" in the UI:

1. `_FixCodeProcessor.runFirstPhase`, (the `NonNullableFix.prepareUnit`s), **33 seconds**
2. `_FixCodeProcessor.runLaterPhases`, (the `NonNullableFix.processUnit`s), **53 seconds**
3. `_FixCodeProcessor.runLaterPhases`, (the `NonNullableFix.finalizeUnit`s), **42 seconds**
4. `_FixCodeProcessor.runLaterPhases`, (`NonNullableFix.finish` ,including `InfoBuilder.explainMigration`), **330 seconds**
5. `NonNullableFix.rerun`, (`MigrationState.refresh`, which also calls `InfoBuilder.explainMigration`), **387 seconds**

### Third "Rerun" in the UI:

1. `_FixCodeProcessor.runFirstPhase`, (the `NonNullableFix.prepareUnit`s), **35 seconds**
2. `_FixCodeProcessor.runLaterPhases`, (the `NonNullableFix.processUnit`s), **56 seconds**
3. `_FixCodeProcessor.runLaterPhases`, (the `NonNullableFix.finalizeUnit`s), **41 seconds**
4. `_FixCodeProcessor.runLaterPhases`, (`NonNullableFix.finish` ,including `InfoBuilder.explainMigration`), **278 seconds**
5. `NonNullableFix.rerun`, (`MigrationState.refresh`, which also calls `InfoBuilder.explainMigration`), **356 seconds**
